[10.0][FIX] purchase_packaging: purchase line product qty should be s…

rousseldenis opened the pull request:
…tored

rousseldenis committed:
[10.0][FIX] purchase_packaging: purchase line product qty should be stored

rousseldenis commented:
@Cedric-Pigeon

pedrobaeza commented:
If you have an inverse, I think you have to change the compute method for taking that store value for being consistent.

Cedric-Pigeon commented on purchase_packaging/migrations/10.0.1.0.5/post-migrate.py:
```diff
@@ -0,0 +1,10 @@
+import odoo
```
This file should be named post-migration.py

Cedric-Pigeon replied:
Sorry post-migrate is also valid.

rousseldenis commented:
> If you have an inverse, I think you have to change the compute method for taking that store value for being consistent.

I re read this and there is no change in the logic of compute/inverse here. Just store the value.

I will explore travis errors

rousseldenis committed:
[10.0][FIX] purchase_packaging: Recompute product_qty

If product_qty is contained in dict for create() and write(),
it is written unchanged in DB.
Change that behaviour and recompute it once.

rousseldenis commented:
This one is greenifying!

This change is quite wanted as the product_qty field on order lines was computed one not stored. That would be annoying if we want to print some reports as the value was void.

As the tests lines amount is quite covering much cases (and are not changed here), we would think this is acceptable.

rousseldenis commented:
@Cedric-Pigeon Could you udpate your review ?

rousseldenis commented:
@lmignon

lmignon commented on purchase_packaging/models/purchase_order_line.py:
```diff
@@ -186,8 +187,20 @@ def create(self, vals):
+        # environment that would help distinguish write call from a
+        # compute
+        if 'product_qty' in vals and not \
+                self.env.context.get('no_recompute_product_qty'):
```
```suggestion
                self.env.context.get('recomputing_product_qty'):
```

lmignon commented on purchase_packaging/models/purchase_order_line.py:
```diff
@@ -186,8 +187,20 @@ def create(self, vals):
+        if 'product_qty' in vals and not \
+                self.env.context.get('no_recompute_product_qty'):
+            self.with_context(
+                no_recompute_product_qty=True)._compute_product_qty()
```
```suggestion
                recomputing_product_qty=True)._compute_product_qty()
```

rousseldenis replied:
Done!

lmignon reviewed (changes requested):
@rousseldenis  What do you think about my naming proposal?

rousseldenis committed:
[10.0][IMP] purchase_packaging: Add context test for create

Avoid a compute call after a create.
Add tests for stored values.

rousseldenis commented:
@lmignon Could you update your review ? This one seems ready

Cedric-Pigeon commented on purchase_packaging/migrations/10.0.1.0.8/post-migrate.py:
```diff
@@ -0,0 +1,9 @@
+    env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+    void_purchase_lines = env['purchase.order.line'].search([
+        ('product_qty', '=', False)])
+    void_purchase_lines._compute_product_qty()
```
you have to set this value in context to prevent recursivity: with_context(recomputing_product_qty=True)
see: https://github.com/OCA/stock-logistics-warehouse/pull/493/files#diff-0b094619c99f4fabbe358caff73c8626R205

rousseldenis committed:
[10.0][FIX] purchase_packaging: Better product_qty migration

sbidoul commented:
@rousseldenis @lmignon @Cedric-Pigeon is this one ready to merge?

lmignon commented:
@sbidoul Yes it's ready.

sbidoul commented:
/ocabot merge minor
